nni/nas/strategy/rl.py

```diff
@@ -32,47 +38,197 @@ class PolicyBasedRL(BaseStrategy):
+        self._replay_buffer: ReplayBuffer | None = None
+
+    def extra_repr(self) -> str:
+        return f'samples_per_update={self.samples_per_update}, replay_buffer_size={self.replay_buffer_size}, ' + \
+            f'reward_for_invalid={self.reward_for_invalid}'
+
+    def _harvest_running_models(self) -> bool:
+        """Harvest completed models and add their trajectories to replay buffer.
+
+        Return true if the policy has just been updated on the latest buffer or no new trajectories found.
+        False otherwise.
+        """
+        running_indices = []
+        recently_updated = True
+        for index, (model, generator) in enumerate(self._running_models):
+            if model.status.completed():
+                if model.metric is not None:
+                    # No matter success or failure, as long as it is completed and gets a metric.
+                    trajectory = generator.send_reward(model.metric)
+                    _logger.info('[Trajectory %4d] (%s, %s) %s', self._trajectory_count + 1, model.status.value, model.metric, model.sample)
+                    recently_updated = self._add_trajectory(trajectory)
+                else:
+                    _logger.info('%s has no metric. Skip.', model)
+            else:
+                running_indices.append(index)
+
+        self._running_models = [self._running_models[i] for i in running_indices]
+        return recently_updated
+
+    def _add_trajectory(self, trajectory: ReplayBuffer) -> bool:
+        """Add the trajectory to replay buffer and execute update if necessary.
+
+        Return true if an update is just executed. False otherwise.
+        """
+        assert self._replay_buffer is not None
+        self._replay_buffer.update(trajectory)
+        self._trajectory_count += 1
+        if self._trajectory_count % self.samples_per_update == 0:
+            self._update_policy()
+            return True
+        return False
+
+    def _update_policy(self) -> None:
+        """Update the RL policy on current replay buffer."""
+        _logger.info('[Trajectory %4d] Updating policy...', self._trajectory_count)
+        assert self._policy is not None and self._replay_buffer is not None
+        update_times = self.update_kwargs.get('update_times', 1)
+        for _ in range(update_times):
+            self._policy.update(0, self._replay_buffer, **self.update_kwargs)
+
+    def _initialize(self, model_space: ExecutableModelSpace, engine: ExecutionEngine) -> ExecutableModelSpace:
+        generator = TuningTrajectoryGenerator(model_space, self.policy_fn)
+        self._policy = generator.policy
+
+        if self.replay_buffer_size is None:
+            replay_buffer_size = generator.expected_trajectory_length * self.samples_per_update
+        else:
+            replay_buffer_size = self.replay_buffer_size
+
+        self._replay_buffer = ReplayBuffer(replay_buffer_size)
+
+        return model_space
+
+    def _run(self) -> None:
```

Comment:
where is `run()` function?

Reply:
Subclass needs to override `_run` and `_initialize`. `run` and `initialize` is defined in base class. See #5371